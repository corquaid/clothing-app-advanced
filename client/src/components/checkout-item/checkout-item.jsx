import React from 'react';
import { connect } from 'react-redux';

import { clearItemFromCart, addItem, removeItem } from '../../redux/cart/cart.actions';

import './checkout-item.scss';

const CheckoutItem = ({ cartItem, clearItem, addItem, removeItem }) => {
  const { imageUrl, name, price, quantity } = cartItem;

  return (
    <div className='checkout-item'>
      <div className='image-container'>
        <img src={imageUrl} alt='item' />
      </div>
      <span className='name'>{name}</span>
      <span className='quantity'>
        <div className='arrow' onClick={() => removeItem(cartItem)}>&#10094;</div> {/* left-facing decrease arrow */}
        <span className='value'>{quantity}</span>
        <div className='arrow' onClick={() => addItem(cartItem)}>&#10095;</div> {/* right-facing increase arrow */}
      </span>
      <span className='price'>{price}</span>
      <div className='remove-button' onClick={() => clearItem(cartItem)}>&#10005;</div>
    </div>
)};

const mapDispatchToProps = dispatch => ({ // clearItem, addItem, removeItem will now be accessible in props above for button onClick event
  clearItem: item => dispatch(clearItemFromCart(item)),
  addItem: item => dispatch(addItem(item)),
  removeItem: item => dispatch(removeItem(item))
});

export default connect(null, mapDispatchToProps)(CheckoutItem);