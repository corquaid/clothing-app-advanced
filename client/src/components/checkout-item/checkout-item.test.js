import React from "react";
import { shallow } from "enzyme";
import { CheckoutItem } from "./checkout-item";

// multiple tests required so let's group them:

describe("CheckoutItem component testing", () => {
    // define required variables/functions

    let wrapper;
    let mockClearItem;
    let mockAddItem;
    let mockRemoveItem;

    // invoke onclick functions
    beforeEach(() => {
        mockAddItem = jest.fn();
        mockClearItem = jest.fn();
        mockRemoveItem = jest.fn();

        // create general mockProps

        const mockProps = {
            cartItem: {
                imageUrl: "www.testimage.com",
                name: "woolly hat",
                price: 40,
                quantity: 2,
            },
            clearItem: mockClearItem,
            addItem: mockAddItem,
            removeItem: mockRemoveItem,
        };

        // general component render wrapper
        wrapper = shallow(<CheckoutItem {...mockProps} />);
    });

    // Define unit tests

    it('Should render the CheckoutItem component', () => {
        expect(wrapper).toMatchSnapshot();
    })

    it('should call removeItem when left arrow is clicked', () => {
        wrapper.find('.left-arrow').simulate('click');
        expect(mockRemoveItem).toHaveBeenCalled();
    })

    it('should call addItem when right arrow is clicked', () => {
        wrapper.find('.right-arrow').simulate('click');
        expect(mockAddItem).toHaveBeenCalled();
    })

    it('should call clearItem when remove button is clicked', () => {
        wrapper.find('.remove-button').simulate('click')
        expect(mockClearItem).toHaveBeenCalled();
    })
});
