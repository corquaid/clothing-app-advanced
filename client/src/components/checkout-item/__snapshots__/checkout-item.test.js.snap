// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckoutItem component testing Should render the CheckoutItem component 1`] = `ShallowWrapper {}`;
